Sources/ComposableArchitecture/Effects/Throttling.swift

```diff
@@ -19,34 +19,46 @@ extension Effect {
     scheduler: S,
     latest: Bool
   ) -> Effect where S: Scheduler {
-    self.flatMap { value -> AnyPublisher<Output, Failure> in
-      guard let throttleTime = throttleTimes[id] as! S.SchedulerTimeType? else {
-        throttleTimes[id] = scheduler.now
-        throttleValues[id] = nil
-        return Just(value).setFailureType(to: Failure.self).eraseToAnyPublisher()
-      }
+    self.receive(on: scheduler)
+      .flatMap { value -> AnyPublisher<Output, Failure> in
+        throttleLock.lock()
+        defer { throttleLock.unlock() }
```

Comment:
Oh, and one other note, the library has an internal `sync` helper for `NSRecursiveLock` (we use a lock for managing cancellation elsewhere), so you could potentially employ it throughout:

```swift
throttleLock.sync {
  ...
}
```

`defer` is probably straightforward enough, but gives us less granular control.

Reply:
Thanks for the heads up, had missed that nice helper! 🙈  I'll change it so use `sync` 👍🏼

Reply:
However, in the `flatMap` "main" body I think it's perhaps better to leave it as is, because it's easier to ensure no escapes from the critical section.

In the `handleEvents` below the `.sync` makes complete sense and makes it much terser.

What do you say?

Reply:
I think the way you have it is fine! We can always reorganize if we feel any reason to later. Thanks :smile: